develop/man/man3/programming.3
.TH "programming" 3 "Fri Oct 7 2016" "Version Version 2.8.1" "WiredTiger" \" -*- nroff -*-
.ad l
.nh
.SH NAME
programming \- Writing WiredTiger applications 
This section covers topics of interest for programmers writing WiredTiger applications\&.
.PP
We follow SQL terminology: a database is set of tables managed together\&. Tables consist of rows, where each row is a key and its associated value\&. Tables may optionally have an associated schema, splitting the value into a set of columns\&. Tables may also have associated indices, each of which is ordered by one or more columns\&.
.PP
.SS "Using the API"
.PP
.IP "\(bu" 2
\fBGetting Started with the API\fP
.IP "\(bu" 2
\fBConfiguration Strings\fP
.IP "\(bu" 2
\fBCursors\fP
.IP "\(bu" 2
\fBTransactions\fP
.IP "\(bu" 2
\fBError handling\fP
.PP
.PP
.SS "Storage options"
.PP
.IP "\(bu" 2
\fBSchema, Columns, Column Groups, Indices and Projections\fP
.IP "\(bu" 2
\fBLog-Structured Merge Trees\fP
.IP "\(bu" 2
\fBFile formats and compression\fP
.IP "\(bu" 2
\fBCompressors\fP
.IP "\(bu" 2
\fBEncryptors\fP
.PP
.PP
.SS "Programming notes"
.PP
.IP "\(bu" 2
\fBMultithreading\fP
.IP "\(bu" 2
\fBName spaces\fP
.IP "\(bu" 2
\fBDatabase read-only mode\fP
.IP "\(bu" 2
\fBSignal handling\fP
.PP
.PP
.SS "Advanced topics"
.PP
.IP "\(bu" 2
\fBAsynchronous operations\fP
.IP "\(bu" 2
\fBBackups\fP
.IP "\(bu" 2
\fBCompaction\fP
.IP "\(bu" 2
\fBCheckpoint durability\fP
.IP "\(bu" 2
\fBCommit-level durability\fP
.IP "\(bu" 2
\fBIn-memory databases\fP
.IP "\(bu" 2
\fBJoin cursors\fP
.IP "\(bu" 2
\fBLog cursors\fP
.IP "\(bu" 2
\fBNamed Snapshots\fP
.IP "\(bu" 2
\fBRebalance\fP
.IP "\(bu" 2
\fBPer-process shared caches\fP
.IP "\(bu" 2
\fBStatistics\fP
.IP "\(bu" 2
\fBUpgrading and downgrading databases\fP
.PP
.PP
.SS "Extending WiredTiger"
.PP
.IP "\(bu" 2
\fBExtending WiredTiger\fP
.IP "\(bu" 2
\fBCustom Collators\fP
.IP "\(bu" 2
\fBCustom Extractors\fP
.IP "\(bu" 2
\fBCustom Data Sources\fP
.IP "\(bu" 2
\fBCustom File Systems\fP
.IP "\(bu" 2
\fBWiredTiger Helium support\fP
.PP
.PP
.SS "Performance monitoring and tuning"
.PP
.IP "\(bu" 2
\fBPerformance monitoring with statistics\fP
.IP "\(bu" 2
\fBSimulating workloads with wtperf\fP
.IP "\(bu" 2
\fBVisualizing performance with wtstats\fP 
.PP
.PP
.IP "\(bu" 2
\fBMemory allocator\fP
.IP "\(bu" 2
\fBPage and overflow key/value sizes\fP
.IP "\(bu" 2
\fBCache and eviction tuning\fP
.IP "\(bu" 2
\fBBulk-load\fP
.IP "\(bu" 2
\fBCursor persistence\fP
.IP "\(bu" 2
\fBRead-only objects\fP
.IP "\(bu" 2
\fBCommit-level durability\fP
.IP "\(bu" 2
\fBChecksums\fP
.IP "\(bu" 2
\fBCompression\fP
.IP "\(bu" 2
\fBFile allocation\fP
.IP "\(bu" 2
\fBSystem buffer cache\fP
.IP "\(bu" 2
\fBLinux transparent huge pages\fP
.IP "\(bu" 2
\fBConnection close\fP
.IP "\(bu" 2
\fBMutexes\fP
.IP "\(bu" 2
\fBLinux zone reclamation memory management\fP 
.PP



1.4.2/programming.js
var programming =
[
    [ "Getting Started with the API", "basic_api.html", [
      [ "Connecting to a database", "basic_api.html#basic_connection", null ],
      [ "Creating a table", "basic_api.html#basic_create_table", null ],
      [ "Accessing data with cursors", "basic_api.html#basic_cursors", null ],
      [ "Closing handles", "basic_api.html#basic_close", null ]
    ] ],
    [ "Configuration Strings", "config_strings.html", [
      [ "Introduction", "config_strings.html#config_intro", null ],
      [ "JavaScript Object Notation (JSON) compatibility", "config_strings.html#config_json", null ]
    ] ],
    [ "Error handling", "error_handling.html", null ],
    [ "Cursors", "cursors.html", "cursors" ],
    [ "Multithreading", "threads.html", [
      [ "Code samples", "threads.html#threads_example", null ]
    ] ],
    [ "Schemas", "schema.html", "schema" ],
    [ "File formats and compression", "file_formats.html", "file_formats" ],
    [ "Log-Structured Merge Trees", "lsm.html", [
      [ "Background", "lsm.html#lsm_background", null ],
      [ "Description of LSM trees", "lsm.html#lsm_description", null ],
      [ "Interface to LSM trees", "lsm.html#lsm_api", null ],
      [ "Merging", "lsm.html#lsm_merge", null ],
      [ "Bloom filters", "lsm.html#lsm_bloom", null ],
      [ "Creating tables using LSM trees", "lsm.html#lsm_schema", null ],
      [ "Caveats", "lsm.html#lsm_caveats", [
        [ "Hazard configuration", "lsm.html#lsm_hazard", null ],
        [ "Empty values", "lsm.html#lsm_tombstones", null ],
        [ "Transactional access", "lsm.html#lsm_txn", null ]
      ] ]
    ] ],
    [ "Transactions", "transactions.html", [
      [ "ACID properties", "transactions.html#transactions_acid", null ],
      [ "Transactional API", "transactions.html#transactions_api", null ],
      [ "Implicit transactions", "transactions.html#transactions_implicit", null ],
      [ "Concurrency control", "transactions.html#transactions_concurrency", null ],
      [ "Isolation levels", "transactions.html#transaction_isolation", null ],
      [ "Checkpoints and Recovery", "transactions.html#transaction_recovery", null ]
    ] ],
    [ "Cache configuration", "cache_configuration.html", [
      [ "Overview of WiredTiger cache configuration.", "cache_configuration.html#cache_basic", null ],
      [ "Overview of WiredTiger shared cache configuration.", "cache_configuration.html#shared_cache", null ],
      [ "Overview of WiredTiger eviction configuration.", "cache_configuration.html#cache_eviction", null ]
    ] ],
    [ "Checkpoints", "checkpoints.html", null ],
    [ "Hot backup", "hot_backup.html", null ],
    [ "Compressors", "compression.html", [
      [ "Using bzip2 compression", "compression.html#compression_bzip2", null ],
      [ "Using snappy compression", "compression.html#compression_snappy", null ],
      [ "Upgrading compression engines", "compression.html#compression_upgrading", null ],
      [ "Custom compression engines", "compression.html#compression_custom", null ]
    ] ],
    [ "Database Home Directory", "home.html", null ],
    [ "Database Configuration", "database_config.html", [
      [ "WiredTiger.config file", "database_config.html#config_file", null ],
      [ "WIREDTIGER_CONFIG environment variable", "database_config.html#config_env", null ]
    ] ],
    [ "Programmatic name spaces", "name_space.html", [
      [ "Process' environment name space", "name_space.html#env", null ],
      [ "C language name space", "name_space.html#c", null ],
      [ "File system name space", "name_space.html#filename", null ],
      [ "Error return name space", "name_space.html#error", null ]
    ] ],
    [ "Security", "security.html", [
      [ "Database directory permissions", "security.html#directory_permissions", null ],
      [ "File permissions", "security.html#file_permissions", null ],
      [ "Environment variables", "security.html#environment_variables", null ]
    ] ],
    [ "Signal handling", "signals.html", null ],
    [ "Performance Tuning", "tuning.html", [
      [ "Cache size", "tuning.html#tuning_cache_size", null ],
      [ "Cache resident objects", "tuning.html#tuning_cache_resident", null ],
      [ "Page and overflow sizes", "tuning.html#tuning_page_size", null ],
      [ "Memory allocation", "tuning.html#tuning_memory_allocation", null ],
      [ "Checksums", "tuning.html#tuning_checksums", null ],
      [ "Performance monitoring with statistics", "tuning.html#tuning_statistics", null ]
    ] ]
];